odbc/src/api/driver_connect_tests.rs

```diff
@@ -0,0 +1,142 @@
+        #[test]
+        fn $func_name() {
+            use odbc_sys::SmallInt;
+            let in_connection_string = $in_connection_string;
+            let driver_completion = $driver_completion;
+            let expected_sql_state = $expected_sql_state;
+            let expected_sql_return = $expected_sql_return;
+            let expected_error_message = $expected_error_message;
+
+            let out_connection_string = &mut [0u16; 64] as *mut _;
+            let string_length_2 = &mut 0;
+            let buffer_length: SmallInt = 65;
+            let env_handle: *mut MongoHandle =
+                &mut MongoHandle::Env(RwLock::new(Env::with_state(EnvState::Allocated)));
+            let conn_handle: *mut MongoHandle = &mut MongoHandle::Connection(RwLock::new(
+                Connection::with_state(env_handle, ConnectionState::Allocated),
+            ));
+            let mut expected_sql_state_encoded: Vec<u16> =
+                expected_sql_state.encode_utf16().collect();
+            expected_sql_state_encoded.push(0);
+            let mut in_connection_string_encoded: Vec<u16> =
+                in_connection_string.encode_utf16().collect();
+            in_connection_string_encoded.push(0);
+
+            let actual_return_val = SQLDriverConnectW(
+                conn_handle as *mut _,
+                std::ptr::null_mut(),
+                in_connection_string_encoded.as_ptr(),
+                in_connection_string.len().try_into().unwrap(),
+                out_connection_string,
+                buffer_length,
+                string_length_2,
+                driver_completion,
+            );
+            assert_eq!(expected_sql_return, actual_return_val);
+
+            let text_length_ptr = &mut 0;
+            let actual_sql_state = &mut [0u16; 6] as *mut _;
+            let actual_message_text = &mut [0u16; 512] as *mut _;
+            // Using SQLGetDiagRecW to get the sql state and error message
+            // from the connection handle
+            let _ = SQLGetDiagRecW(
+                HandleType::Dbc,
+                conn_handle as *mut _,
+                1,
+                actual_sql_state,
+                &mut 0,
+                actual_message_text,
+                512,
+                text_length_ptr,
+            );
+            let actual_message_length = *text_length_ptr as usize;
+            unsafe {
+                assert_eq!(
+                    expected_error_message,
+                    &(String::from_utf16_lossy(&*(actual_message_text as *const [u16; 256])))
+                        [0..actual_message_length],
+                );
+                assert_eq!(
+                    String::from_utf16(&*(expected_sql_state_encoded.as_ptr() as *const [u16; 6]))
+                        .unwrap(),
+                    String::from_utf16(&*(actual_sql_state as *const [u16; 6])).unwrap()
+                );
+            }
+        }
+    };
+}
+
+mod unit {
```

Comment:
I am actually not sure that these can really be called unit tests, it sounds more like integration tests. Maybe they should move to /odbc/tests/connection_tests.rs?
And the part that checks the diagnostics could go into the odbc/tests/common/mod.rs.
Unit tests could be added, if you think it's necessary, targeting sql_driver_connect and simply checking if the result is OK or Error.
Looks like this is something for [SQL-1036](https://jira.mongodb.org/browse/SQL-1036).
WDYT?

Reply:
That sounds reasonable to me.  
I'll add these steps to SQL-1036 and we can update the testing at that point.